### PR TITLE
darkroom : show toast message on rating & colorlabels

### LINKED ISSUE #6037: indicate the success of rejecting an image in darkroom when "all except rejected" filter is active
Reference #5968 

Please return the indication of success when rejecting an image in darkroom when the view filter is set to "all except rejected".

ideally should advance to next image in selection and set that image as selected.

One should NEVER be able to perform work on an image that is not "selected"!

### PR DESCRIPTION
**Pull request by AlicVB**
this fix #6037

toast msg appears only in darkroom if the impacted image is the current one (this doesn't affect filmstrip)
This is needed because otherwise we have no indication of rating/colorlabels change

To be able to show colorlabels in... colors, I've updated toast msg to allow markup. This shouldn't change "normal" text

### CODE REVIEW

File: src/dtgtk/thumbtable.c [hunk 1]
```diff
@@ -2017,6 +2042,54 @@ static gboolean _accel_color(GtkAccelGroup *accel_group, GObject *acceleratable,
...
+            result = dt_util_dstrcat(result, "<span foreground=\"#00ee00\">⚫ </span>");
+          }
+          else if(g_strcmp0(lb, "blue") == 0)
+          {
+            result = dt_util_dstrcat(result, "<span foreground=\"#0000ee\">⚫ </span>");
+          }
+          else if(g_strcmp0(lb, "purple") == 0)
+          {
+            result = dt_util_dstrcat(result, "<span foreground=\"#ee00ee\">⚫ </span>");
+          }
```
Comment (johnny-bit) [inline src/dtgtk/thumbtable.c:2081]: This is great, but there's a problem we've dealt with back when similar code was used for info line. Basically if user has colormoji font (or something like that), they'll see just  "colorlabels set to :black_circle: :black_circle: :black_circle: ".

Maybe add color name in parenthesis for those?
eg: "colorlabels set to :black_circle: (red) :black_circle: (green) :black_circle: (blue)".

for people who don't have colormoji font that would look like:
"colorlabels set to :red_circle: (red) :green_circle: (green) :large_blue_circle: (blue)".

Comment (TurboGit) [reply src/dtgtk/thumbtable.c:2081]: Or use a standard symbol char like : ●

Comment (TurboGit) [reply src/dtgtk/thumbtable.c:2081]: Or a larger version ⬤

The point is that using standard symbols will work in all cases.

**Review by TurboGit — CHANGES_REQUESTED** [review]
Use a standard symbol character.

**Comment by dterrahe** [comment]
> To be able to show colorlabels in... colors, I've updated toast msg to allow markup. This shouldn't change "normal" text

In theory, you should escape "normal" text using g_markup_escape_text on the off chance that they contain <>& characters. As could be the case if anybody adds a module or bauhaus widget with such a character in its name that would then have dynamic accelerators.

Maybe rename dt_toast_log -> dt_toast_markup_log and introduce a new dt_toast_log that just calls dt_toast_markup_log(g_markup_escape_text(), ...).

File: src/dtgtk/thumbtable.c [hunk 2]
```diff
@@ -2009,6 +2010,30 @@ static gboolean _accel_rate(GtkAccelGroup *accel_group, GObject *acceleratable,
...
+      if(img)
+      {
+        const int r = img->flags & DT_IMAGE_REJECTED ? DT_VIEW_REJECT : (img->flags & DT_VIEW_RATINGS_MASK);
+        dt_image_cache_read_release(darktable.image_cache, img);
+
+        // translate in human readable value
+        if(r == DT_VIEW_REJECT)
+          dt_toast_log(_("image rejected"));
+        else
+          dt_toast_log(ngettext("image rated to %d star", "image rated to %d stars", r), r);
```
Comment (TurboGit) [inline src/dtgtk/thumbtable.c:2032]: BTW, we would also display a number of ★ corresponding to the rating. Should be easy with a simple table of strings.

**Comment by AlicVB** [comment]
- I've reworked the markup part with toast so we now can choose to use markup or not when displaying msg. @dterrahe : I've not done that exactly as you suggest, because it was difficult to avoid code duplication with the special definition of this fct...
- I've change the bullets with @TurboGit char suggestion.
- Same for stars

**Comment by TurboGit** [comment]
Works for me. Merging to master & queued for 3.4.1. Thanks.

**Comment by dterrahe** [comment]
> * I've reworked the markup part with toast so we now can choose to use markup or not when displaying msg. @dterrahe : I've not done that exactly as you suggest, because it was difficult to avoid code duplication with the special definition of this fct...

After a bit of research, the correct way seems to be to rework the existing function to take a va_list argument and the boolean you introduced (call it _toast_log_va), and then have toast_log and toast_markup_log call it after setting up the va_list from their respective ... arguments.
```
va_list ap;
va_start(ap, msg);
_toast_log_va(FALSE/TRUE, msg, ap);
va_end(ap);
```
So only 3 lines are duplicated (and an extra function call).

I'd prefer this solution, because in my experience these code snippets will get copied all over the place. Always using TRUE _seems_ to work, because people won't be aware of the corner cases and therefore not test them.

**Comment by dterrahe** [comment]
> Works for me. Merging to master & queued for 3.4.1. Thanks.

You are leaving _very_ little time for review here! ;-)

**Comment by TurboGit** [comment]
> You are leaving _very_ little time for review here! ;-)

Yes, sorry :( I try to clean-up the pile of PR before it starts being unmaintainable on my side. I'll try to let more time.

**Comment by dterrahe** [comment]
No worries, I was joking. I thought you would take a rest after 3.4 was tagged, but it looks like the opposite...

We have a whole year for 3.6; sometimes it is not bad to let a patch "mature"; otherwise you might get patches on top of patches and end up doing more work on your side.

But it is also very satisfying to see all the little fixes and improvements like this!

**Comment by AlicVB** [comment]
@dterrahe : as we need another PR for what you propose, can I let you do it ? I'm not sure to understand exactly what you means...

**Comment by TurboGit** [comment]
> But it is also very satisfying to see all the little fixes and improvements like this!

Right, and I'm more in a mode "do incremental fix/enhancement and commit often" because this lets more time for field testing. We sadly not have automatic test for the GUI and it becomes so complex that it is almost impossible to know what kind of regression or unexpected side effect some commits may have. So when I have a PR at hand, that the code looks ok and if it is working I do commit. If we can improve fine, let's do another PR. That being said you had reacted to the PR so it would have been nice to wait for you to come back to it.
